### DISCUSSION TITLE
Using images/graphics: develop guidelines

### DISCUSSION THREAD

**Discussion by lena-voita** [machinetranslate/machinetranslate.org#234] [General]
## I believe we need graphics
One of the main goals of the website is to express high-level ideas very clearly to make people understand the content easily (https://machinetranslate.org/style). I believe good graphics are one of the ways to achieve this: in many cases, information presented visually (when done properly) is much easier to perceive/comprehend than reading text. I’m afraid that forbidding graphics would harm the website and make achieving our goal harder. Below I'll show some of the examples where I think graphics could help to illustrate the point and show what I have in mind.

## But there are some concerns
Of course, there are some concerns regarding using graphics which Adam mentioned:
- hard to translate
(content of machinetranslate.org will be machine translated, and graphics are hard to translate)
- hard to edit
- have to keep in mind how this'll look on mobile devices

## Action point: discuss and agree on the guidelines
While I totally agree with the concerns, I still think that forbidding the graphics completely would make it much harder to achieve our goal. What I think we could do:
1. categorise the types of graphics that could be helpful for the site and give examples. I'll do this below and you can add other types/examples 
2. propose some ways to address the concerns. I'll propose something and then we can start discussion from there
3. decide which types of graphics are allowed or not, and for those allowed which guidelines they have to follow

**Comment by lena-voita** [comment]
# Types of graphics and use cases

## Showing structure/logical order
### Elements: mostly text and arrows

Such structures can help navigating and show the connections between different sections. For example, when talking about Quality evaluation, the figure below shows (1) the main difference between quality estimation and evaluation, hence (2) its place in the hierarchy of all the metrics and (3) shows the categorisation of the metrics inside the part we are going to talk about. With this, the structure is clear and the rest is much easier to comprehend. Without this, explaining the structure with plain text would require a person to read lot of text and build it in their head: create all the logical connections etc. I really think having such structures at the top of some of the pages would make it much more clear and would help a lot.

![image](https://user-images.githubusercontent.com/16096019/163408276-b79ed11c-3c0f-46f7-9d4c-81a41c966caa.png)


## Schemes and animations showing how things work

In lots of cases, showing a clear picture/scheme/animation can save a lot of trouble trying to explain the same in text. And, after looking at the picture, it’ll be much easier for a person to comprehend the text explaining the same.

For example, we can illustrate the idea of beam search (just an example - can be updated!).
![beam_search](https://user-images.githubusercontent.com/16096019/163408380-ae7ecb90-da36-4c91-a442-41688d05adc3.gif)

Or show how neural MT works: reads the input sentence with an encoder and and generates output one token at a time. The picture below can be simplified to be adapted for the site (e.g., remove the distributions and show only encoder, decoder and generating word by word).
![enc_dec_prob_idea](https://user-images.githubusercontent.com/16096019/163408522-b3883709-ae5c-486b-8639-3aab76adabd7.gif)

When talking about attention, we can also give an idea of what it is.
![image](https://user-images.githubusercontent.com/16096019/163408618-f710d509-ac62-46e4-9bf1-59377dca26e7.png)

Or what a model learns: showing an attention map to give an idea of how it works when trained.
![image](https://user-images.githubusercontent.com/16096019/163408582-9920afc3-ee86-489c-bc0d-088b8c4630c1.png)

The idea behind self-attention in Transformers:
![encoder_self_attention](https://user-images.githubusercontent.com/16096019/163408692-d5355466-ecec-4d8b-b58a-c3df608e521e.gif)

## Less formal: sketches giving intuition

This is the least formal type of figures, but sometimes they can be a good way to show the intuition behind some of the stuff. I don’t have yet anything for this project specifically, but below are some examples I did for other topics.

For example, when talking about evaluation of language models, perplexity might be hard for non-technical people. But the intuition behind can be made clear.
<img width="1096" alt="idea" src="https://user-images.githubusercontent.com/16096019/163408736-e7ace6d6-2f9f-4d17-9940-8c4a0873c6e9.png">

Or, when talking about transition from word-based representations to contextualized models, we can illustrate what different embeddings see in the data.
<img width="1238" alt="words_to_context" src="https://user-images.githubusercontent.com/16096019/163408787-bb19d32b-d21a-485f-8521-f53e524c4e45.png">

**Comment by lena-voita** [comment]
## Three types of graphics
So far we have three types of graphics:
(1) schemes showing structures (mostly branching trees)
(2) pictures/animations showing how things work
(3) very informal sketches showing intuition behind technical stuff

## Discuss concerns
Now let’s look at the concerns

### Hard to edit
 - types 2 and 3 (pictures) - not a problem since they are static
 - type 1 (structures): might need editing
#### Ways to alleviate:
 - put there only structure that is fairly static (e.g., in the example with Quality evaluation give the structure but not mention specific metrics that can change)
 - along with the picture, put the source (e.g., .pptx) to let others update if needed

### Hard to translate
 - types 2 and 3 (pictures) - contain a couple of words, but maybe that’s ok. For example, Wikipedia articles also contain schemes and animations that do not change when switching to the same page in different languages.
 - type 1 (structures) - more problematic, but maybe this is also not a big problem.
#### Ways to alleviate:
Since each figure illustrates some idea, the idea itself is discussed in the text. In this sense, figures help but if they are not translated, this does not stop from understanding the content.

### Mobile devices.
We started discussing this here: https://github.com/machinetranslate/machinetranslate.org/pull/228

Summing up, Cecilia mentioned that the article content section has a somewhat fixed width, and anything that may exceed that width would require readers to left-right scroll. This would also affect phone navigability. I replied that what we could do is just set image width in percentage of the screen width rather than a fixed absolute width value - can we? In html, this works well. 
**Does anyone know if this can be done?**

# Action point: let's discuss!

I need your help to agree on the following:
1. for each of the graphic types, whether they are allowed or not
2. if allowed, what are the guidelines to alleviate the concerns

**Comment by bittlingmayer** [comment]
Fixed width or max width images - it can be done, we do it I think, but they're very small on mobile.

So we should try to make images more tall than wide, and not too detailed.

**Comment by bittlingmayer** [comment]
One way out of this is to use graphics from elsewhere and just clearly say that.

That's what I did on https://machinetranslate.org/hybrid-translation.

Then we don't need to worry about a consistent style, just like we don't worry that paper titles don't follow our style guide.

The examples you gave of type 2 and 3 would be covered by that.

e.g. we can describe it a slide from a 2020 presentation by Lena Voita.

Machine Translate is intended to be meta, after all.

**Comment by bittlingmayer** [comment]
Type 1 I am less sure about.

At least that example is a DAG so it can be perfectly represented with a Markdown list or YAML list with some nesting.

It's also text-heavy, so having it in Markdown means better translation, page search, site search, web search, better sizing on mobile...

(Whereas text input examples in a transformer example don't need to be indexed or translated.)

There are ways to do diagrams in HTML/CSS, but not sure it's necessary in this case.

In many cases, hierarchies like in type 1 can be implicit in the site/section/page nav.

**Comment by cefoo** [comment]
For the first graphic, we may use something like this:

---

<h2>Does the metric use reference?</h2>

<h3>- Quality evaluation</h3>
(A good metric correlates well with human judgements)

- **Similarity** between machine and human translations

     - N-gram matching
     - Embeddings similarity
       
- **Correlation** with human judgement (optimized directly)
    
<h3>- Quality estimation</h3>
...
